File: 54.spiralOrder.c
#include "shdstd.h"

int *spiralOrder(int **matrix, int size, int *col_size, int *ret_size)
{
	int rows = size;
	int cols = col_size[0];

	--rows;

	int lim = cols;

	int R = 0;
	int C = -1;

	*ret_size = size * (col_size[0]);
	int *result = malloc((*ret_size) * sizeof(int));
	int w = 0;

	int dr = 0;
	int dc = 1;

	for (;;) {
		if (lim <= 0) {
			break;
		}

		for (int i = 0; i < lim; ++i) {
			R += dr;
			C += dc;

			result[w++] = matrix[R][C];
		}

		if (dr != 0) {
			--rows;
			--cols;
			lim = cols;
		} else {
			lim = rows;
		}

		int tmp = dc;
		dc = -(dr);
		dr = tmp;
	}

	return result;
}

File: 54.spiralOrder.go
package main

func spiralOrder(matrix [][]int) []int {
	rows := len(matrix)
	cols := len(matrix[0])

	rows--
	lim := cols

	R := 0
	C := -1

	resSize := len(matrix) * len(matrix[0])
	result := make([]int, resSize)
	w := 0

	dr := 0
	dc := 1

	for {
		if lim <= 0 {
			break
		}

		for i := 0; i < lim; i++ {
			R += dr
			C += dc
			result[w] = matrix[R][C]
			w++
		}

		if dr != 0 {
			rows--
			cols--
			lim = cols
		} else {
			lim = rows
		}

		tmp := dc
		dc = -(dr)
		dr = tmp
	}

	return result
}
